Add ast as string

mstemm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. Please add a release note!
4. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

> /kind bug

> /kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area build

> /area driver-kmod

> /area driver-bpf

> /area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-udig

> /area libscap

/area libsinsp

/area tests

> /area proposals

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:

Add the ability to print an AST as a string. This is useful in code that works with ASTs that wants to include the string representation in log messages/return values.

**Which issue(s) this PR fixes**:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
If PR is `kind/failing-tests` please post the related issues/tests in a comment and do not use `Fixes`.
-->

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

<!--
If no, you have to do nothing.
If yes, a release note is required:
Delete `NONE` and enter your extended release note in the block below.
Please note, the release note follows the "conventional commit specification" (https://www.conventionalcommits.org/en/v1.0.0/):
For example: `fix: broken link`.
If the PR requires additional action from users switching to the new release, prepend the string "action required:".
For example, `action required: change the API interface of libscap`.
-->

```release-note
NONE
```

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/394#" title="Author self-approved">mstemm</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [mstemm]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

jasondellaluce reviewed (commented):
This is great, I think we definitely need something like this (linked issue: https://github.com/falcosecurity/falco/issues/1814). Thanks Mark!

I have four points:

1. **Nitpicking**: Here we seem to add a space at every `append()`, which results in much spaced strings such as `( proc.name = nginx and fd.name = /etc/passwd )`. I understand this is for simplicity, but I think we can attain something better (e.g. non leaving a space after a `(`) which could help readability. I think this would be a better addition for future use cases. This can be done by adding spaces selectively in every `visit()` case.

2. **Testing**: I LOVE that you contributed with more tests! I just wanted to suggest that another solid testing approach would be to parse a filter to an AST, formatting it as a string, then re-parsing it as an AST and see if the two AST match. In this way we can prove that the AST <-> string is really reversible and bi-directional. Example:
```
TEST(string_visitor, some_check)
{
    std::string in = <some complex filter>;
    std::shared_ptr<ast::expr> e1(parser(in).parse());
    std::shared_ptr<ast::expr> e2(parser(as_string(e1)).parse());
    ASSERT_TRUE(ast::is_equal(e1, e2));
}
```

3. **Memory**: As showed in the code example above, you could use shared pointers to ensure that heap-allocated objects like ASTs are correctly de-allocated during testing. I don't know what are the memory guarantees of gtest, but this is definitely a safer approach to ensure the test suite stays optimized as we it grows.

4.  **String escaping**: At first glance, it looks like you didn't handle string escaping here. My bet is that a test like the one suggested in point (2) would fail if a string value contains spaces or escaped characters. You can look at how strings are unescaped right in the parser implementation (https://github.com/falcosecurity/libs/blob/afca72cec1994a5834f9ec2e9687b4be4a1c3c61/userspace/libsinsp/filter/parser.cpp#L594). I would think that the escape-unescape transformation could become two public static methods of the parser itself! String escaping applies to: field arguments, check right-hand values, and list values.

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/libs/blob/master/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [2c8ef42](https://github.com/falcosecurity/libs/commits/2c8ef427a78eadb512f0b86ca5ba295b895f6069) Squash w Add a ast::as_string() method, using a string visitor.
- [aa7aef4](https://github.com/falcosecurity/libs/commits/aa7aef4dafff5299f639e47666393f9edd688c41) Squash w Add unit tests for ast::as_string

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
@mstemm: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-minimal | 90be8a96e7e41c364e574e00f5c628a495d5eb69 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-minimal/1537620188118126592) | true | `/test build-libs-minimal`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mstemm commented:
Ok addressed all of the above:

1. the string visitor is more selective about adding spaces. The string representation generally always matches the input other than adding parentheses around top level logical operators and always separating a field, op, and values with spaces.
2. The tests have variants that check going from an ast to string back to ast, and comparing the asts. These don't replace the existing tests as you could have trivially wrong asts from the parser that wouldn't be detected if you just compared the asts.
3.  The ast::expr pointer is wrapped in a unique_ptr now. (And a helper method removed the duplicate test code).
4. I added string escaping, added new tests that check string representations of filters with escaped list values, field args, and binary op values. I also added a few unit tests specifically for string escaping + unescaping.

poiana commented:
@mstemm: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-minimal | 90be8a96e7e41c364e574e00f5c628a495d5eb69 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-minimal/1537620188118126592) | true | `/test build-libs-minimal`
build-libs | 90be8a96e7e41c364e574e00f5c628a495d5eb69 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs/1537620187652558848) | true | `/test build-libs`
build-libs-arm | 90be8a96e7e41c364e574e00f5c628a495d5eb69 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1537620187728056320) | true | `/test build-libs-arm`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mstemm commented:
/retest

jasondellaluce commented on userspace/libsinsp/filter/ast.h:
```diff
@@ -26,6 +26,14 @@ namespace libsinsp {
+	\note Throws a sinsp_exception in case of parsing errors.
+	\return an escaped/unescaped verison of the string
+*/
+std::string escape_str(const std::string& str);
```
I would prefer having these as static methods of the `libsinsp::filter::parser` class. Effectively, this logic is dictated by the parsing grammar and is not tied to the AST structures. This would also have the parsing logic all in the same place.

mstemm replied:
I was running into circular dependency problems with that, as ast was being included by parser. But I will declare standalone stuff to make it work.

jasondellaluce replied:
Oh I didn't think of that. In that case creating separate .h and .cpp files could be a solution maybe.

mstemm replied:
I just declared escape_str standalone in ast.cpp.

jasondellaluce commented on userspace/libsinsp/filter/ast.cpp:
```diff
@@ -16,9 +16,143 @@ limitations under the License.
+			break;
+		case '\'':
+			should_escape = true;
+			escape_char = '\'';
```
Note, this will not work. There may be a case in which you already escaped a `"` as `\"` before reaching this point, and this would mix `'` escape delimiters with `\"`. Now, I see two solutions to this:
1. **More conservative**: we decide to always have `"` as an escape delimiter and we remove this line 140. In this way, strings are always escaped with `"` which can result in uglier results in some edge case but it's never ambigue.
2. **Less conservative**: we drop the assumption that delimiters `"` and `'` can't be respectively mixed with `\'` and `\"` (enforced at line 74 and 81 here in `unescape_str()`). This would allow you to achieve what you wanted, but has the downside of potentially creating breaking changes in many existing rulesets, and also makes the grammar more loose and error-prone.

Due to the benefits/downsides tradeoffs described above, my vote goes for option 1). WDYT?

mstemm replied:
I'll do the first thing--always escape with double quotes. That will mean that an expression like `proc.name in ('some proc')` will be parsed and printed as `proc.name in (\"some proc\")`. This breaks the assumption that parsed-and-printed result in the same expression, but if parsed again, it *will* result in an equal ast.

jasondellaluce replied:
I think having equal ASTs is the bigger concern here.

jasondellaluce commented on userspace/libsinsp/filter/ast.cpp:
```diff
@@ -16,9 +16,143 @@ limitations under the License.
 
 using namespace libsinsp::filter::ast;
 
+std::string libsinsp::filter::ast::unescape_str(const std::string& str)
```
As for my other comment, I think both `unescape_str()` and `escape_str` should be defined in `parser.cpp`. This incidentally would also save the git history for `unescape_str()` too.

jasondellaluce commented on userspace/libsinsp/filter/ast.cpp:
```diff
@@ -65,6 +199,109 @@ void base_expr_visitor::visit(list_expr* e) { }
+
+void string_visitor::visit(value_expr* e)
+{
+	m_str += ast::escape_str(e->value);
```
Note (nitpicking): macros (a.k.a. identifiers, in the grammar) are represented by `value_expr` nodes too but should not be escaped. You can tell the difference by maintaining a state boolean in the visitor, turning it true if you visited a `binary_check_expr` node before getting here (and turning it off everywhere else).

mstemm replied:
Actually, I'm not sure how a parser would ever generate an ast for a macro reference containing spaces. Both `some macro and proc.name=nginx` and `\"some macro\" and proc.name=nginx` result in parse failures. But I'll still include the handling for it, in case the parser starts handling it.

jasondellaluce commented on userspace/libsinsp/test/string_visitor.ut.cpp:
```diff
@@ -0,0 +1,261 @@
+
+	std::unique_ptr<ast::expr> e(parser.parse());
+
+	ASSERT_STREQ(as_string(*(e.get())).c_str(), "(((evt.type = open or evt.type = openat) and evt.is_open_write = true and fd.typechar = f and fd.num >= 0) and (fd.filename in (.bashrc, .bash_profile, .bash_history, .bash_login, .bash_logout, .inputrc, .profile, .cshrc, .login, .logout, .history, .tcshrc, .cshdirs, .zshenv, .zprofile, .zshrc, .zlogin, .zlogout) or fd.name in (/etc/profile, /etc/bash ...
```
Doesn't `unidirectional()` work here?

mstemm replied:
Oh yes, it would. Fixed.

jasondellaluce commented on userspace/libsinsp/test/string_visitor.ut.cpp:
```diff
@@ -0,0 +1,261 @@
+/*
```
This looks great! I think my only suggestion would be to turn merge all the `TEST_F` for the unidirectional/bidirectional case. In all cases, you could have only one  `TEST_F` that calls `unidirectional()` first and `bidirectional()` after. This is just a minor comment to make the code less verbose, feel free to ignore it if you like it this way!

mstemm replied:
It's a bit of a preference, but I prefer having each TEST/TEST_F only test a single test case, as it's easier to pinpoint the failure when a test fails.

jasondellaluce reviewed (commented):
Thanks for addressing all this Mark! I left you some last concerns and then I think this is good to go!

poiana commented:
@mstemm: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1537802811226460161) | true | `/test build-libs-arm`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@mstemm: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1537802811226460161) | true | `/test build-libs-arm`
build-libs | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs/1537802810073026560) | true | `/test build-libs`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@mstemm: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1537802811226460161) | true | `/test build-libs-arm`
build-libs | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs/1537802810073026560) | true | `/test build-libs`
build-libs-minimal | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-minimal/1537802811226460160) | true | `/test build-libs-minimal`
build-libs-with-chisels | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-with-chisels/1537802811230654464) | true | `/test build-libs-with-chisels`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@mstemm: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1537802811226460161) | true | `/test build-libs-arm`
build-libs | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs/1537802810073026560) | true | `/test build-libs`
build-libs-minimal | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-minimal/1537802811226460160) | true | `/test build-libs-minimal`
build-libs-with-chisels | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-with-chisels/1537802811230654464) | true | `/test build-libs-with-chisels`
build-libs-bundled-deps | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-bundled-deps/1537802811226460162) | true | `/test build-libs-bundled-deps`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@mstemm: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs/1537802810073026560) | true | `/test build-libs`
build-libs-minimal | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-minimal/1537802811226460160) | true | `/test build-libs-minimal`
build-libs-with-chisels | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-with-chisels/1537802811230654464) | true | `/test build-libs-with-chisels`
build-libs-bundled-deps | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-bundled-deps/1537802811226460162) | true | `/test build-libs-bundled-deps`
build-libs-arm | f997fdd3441d5183202718a801b1b0915fbb209a | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1537845885155676160) | true | `/test build-libs-arm`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@mstemm: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-minimal | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-minimal/1537802811226460160) | true | `/test build-libs-minimal`
build-libs-with-chisels | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-with-chisels/1537802811230654464) | true | `/test build-libs-with-chisels`
build-libs-bundled-deps | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-bundled-deps/1537802811226460162) | true | `/test build-libs-bundled-deps`
build-libs-arm | f997fdd3441d5183202718a801b1b0915fbb209a | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1537845885155676160) | true | `/test build-libs-arm`
build-libs | f997fdd3441d5183202718a801b1b0915fbb209a | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs/1537845884308426752) | true | `/test build-libs`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@mstemm: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-with-chisels | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-with-chisels/1537802811230654464) | true | `/test build-libs-with-chisels`
build-libs-bundled-deps | 2260333853fbfa08d77698e54baf8aa4fab9ed3c | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-bundled-deps/1537802811226460162) | true | `/test build-libs-bundled-deps`
build-libs-arm | f997fdd3441d5183202718a801b1b0915fbb209a | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1537845885155676160) | true | `/test build-libs-arm`
build-libs | f997fdd3441d5183202718a801b1b0915fbb209a | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs/1537845884308426752) | true | `/test build-libs`
build-libs-minimal | f997fdd3441d5183202718a801b1b0915fbb209a | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-minimal/1537845885386362880) | true | `/test build-libs-minimal`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

mstemm commented:
/retest

jasondellaluce commented on userspace/libsinsp/filter/parser.cpp:
```diff
@@ -591,7 +591,65 @@ inline const char* parser::cursor()
+			break;
+		case '\'':
+			should_escape = true;
+			res += "\\\"";
```
```suggestion
			res += "'";
```
We should just go for this at this point. `'` is an allowed character if we escape by `"`.
The goal is to enforce the `"` escaping delimiter, but not to change the original string. WDYT?

jasondellaluce commented on userspace/libsinsp/filter/ast.cpp:
```diff
@@ -19,6 +19,17 @@ limitations under the License.
 
 using namespace libsinsp::filter::ast;
 
+// Included in parser.h, not including parser.h here to avoid circular
```
I see why you're doing this, but this can get very fragile. We also have many cases like this in sinsp and this definition-import pattern is getting harder and harder to manage.

At this point, I'd suggest having creating two standalone `escaping.h` and `escaping.cpp`. The `.h` would define:
```
namespace libsinsp {
namespace filter {
std::string escape(const std::string& str);
std::string unescape(const std::string& str);
};
};
```

Then, both AST and parser could import and depend on that. I think this would make this much more manageable and reusable across the codebase. What do you think?

jasondellaluce reviewed (commented):
Left you two more suggestions!

poiana commented:
@mstemm: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1539274369077874688) | true | `/test build-libs-arm`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@mstemm: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1539274369077874688) | true | `/test build-libs-arm`
build-libs-minimal | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-minimal/1539274369207898112) | true | `/test build-libs-minimal`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@mstemm: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1539274369077874688) | true | `/test build-libs-arm`
build-libs-minimal | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-minimal/1539274369207898112) | true | `/test build-libs-minimal`
build-libs | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs/1539274369073680384) | true | `/test build-libs`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@mstemm: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1539274369077874688) | true | `/test build-libs-arm`
build-libs-minimal | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-minimal/1539274369207898112) | true | `/test build-libs-minimal`
build-libs | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs/1539274369073680384) | true | `/test build-libs`
build-libs-with-chisels | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-with-chisels/1539274369161760768) | true | `/test build-libs-with-chisels`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@mstemm: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-arm | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1539274369077874688) | true | `/test build-libs-arm`
build-libs-minimal | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-minimal/1539274369207898112) | true | `/test build-libs-minimal`
build-libs | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs/1539274369073680384) | true | `/test build-libs`
build-libs-with-chisels | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-with-chisels/1539274369161760768) | true | `/test build-libs-with-chisels`
build-libs-bundled-deps | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-bundled-deps/1539274369119817728) | true | `/test build-libs-bundled-deps`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@mstemm: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-minimal | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-minimal/1539274369207898112) | true | `/test build-libs-minimal`
build-libs | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs/1539274369073680384) | true | `/test build-libs`
build-libs-with-chisels | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-with-chisels/1539274369161760768) | true | `/test build-libs-with-chisels`
build-libs-bundled-deps | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-bundled-deps/1539274369119817728) | true | `/test build-libs-bundled-deps`
build-libs-arm | 7e19cbab9eea2fac5436a32e4051054c6ca0e931 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1539275831669428224) | true | `/test build-libs-arm`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

poiana commented:
@mstemm: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
build-libs-minimal | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-minimal/1539274369207898112) | true | `/test build-libs-minimal`
build-libs-with-chisels | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-with-chisels/1539274369161760768) | true | `/test build-libs-with-chisels`
build-libs-bundled-deps | 1693c549147e7ffe3293d479209d20d5ddeb0480 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-bundled-deps/1539274369119817728) | true | `/test build-libs-bundled-deps`
build-libs-arm | 7e19cbab9eea2fac5436a32e4051054c6ca0e931 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs-arm/1539275831669428224) | true | `/test build-libs-arm`
build-libs | 7e19cbab9eea2fac5436a32e4051054c6ca0e931 | [link](http://prow.falco.org/view/s3/falco-prow-logs/pr-logs/pull/falcosecurity_libs/394/build-libs/1539275831585542144) | true | `/test build-libs`

[Full PR test history](http://prow.falco.org/pr-history?org=falcosecurity&repo=libs&pr=394). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/falcosecurity/libs/issues?q=is:issue+is:open) when you hit one in your PR.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

jasondellaluce reviewed (commented):
LGTM! Thanks a lot Mark.

FedeDP commented:
Can you rebase on top of master so that gh actions are run?

FedeDP reviewed (approved):
/approve

poiana commented:
LGTM label has been added.  <details>Git tree hash: f95aff5f3b4eefda2ce14ce5cc89875ca5c87e97</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/libs/pull/394#pullrequestreview-1050693783" title="Approved">FedeDP</a>*, *<a href="https://github.com/falcosecurity/libs/pull/394#" title="Author self-approved">mstemm</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Flibs).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/libs/blob/master/OWNERS)~~ [FedeDP,mstemm]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

Andreagit97 reviewed (approved):
/approve
